onnxruntime/core/providers/mkldnn/nn/conv.cc

```diff
@@ -402,20 +402,30 @@ Status Conv<T>::Compute(OpKernelContext* context) const {
       src_data = static_cast<T*>(dst.get_data_handle());
     }
 
-    // Reorder filter memory layout if necessary.
-    if (filter_format != conv_primitive->GetFilterMemoryFormat()) {
-      auto pd = mkldnn::memory::primitive_desc(mkldnn::memory::desc(filter_dims_mkl,
-                                                                    MklDnnType<T>(),
-                                                                    filter_format),
-                                               cpu_engine);
-      mkldnn::memory src = mkldnn::memory(pd, (void*)filter_data);
-      // allocate the size queried from memory primitive desc. it may not match tensor logical size due to
-      // mkldnn using padding to allow use of blocked format.
-      filter_reorder_buffer = IAllocator::MakeUniquePtr<void>(alloc, conv_primitive->GetFilterSize());
-      mkldnn::memory dst = mkldnn::memory(conv_fwd_pd->weights_primitive_desc(), filter_reorder_buffer.get());
-      MemoryReorderParams params(src, dst);
-      DoReorder<T>(params);
-      filter_data = static_cast<T*>(dst.get_data_handle());
+    // Reorder filter memory layout if necessary
+    // Avoid data reordering. Save filter memory in mkldnn format from first iteration 
+    // in execution provider mapped by weight name.
+    std::string weightName = OpKernel::Node().InputDefs()[1]->Name();
+    std::shared_ptr<mkldnn::memory> filter_dst_mem = nullptr;
+    if(provider_ != nullptr)
+      filter_dst_mem = provider_->GetWeightMemory(weightName);
+
+    if (filter_dst_mem == nullptr) {
+      if (filter_format != conv_primitive->GetFilterMemoryFormat()) {
+        auto pd = mkldnn::memory::primitive_desc(mkldnn::memory::desc(
+          filter_dims_mkl, MklDnnType<T>(), filter_format), cpu_engine);
+        mkldnn::memory src = mkldnn::memory(pd, (void*)filter_data);
+        filter_reorder_buffer = IAllocator::MakeUniquePtr<void>(alloc, conv_primitive->GetFilterSize());
+        filter_dst_mem.reset(
+          new mkldnn::memory(conv_fwd_pd->weights_primitive_desc(), filter_reorder_buffer.get()));
+        MemoryReorderParams params(src, *filter_dst_mem);
+        DoReorder<T>(params);
+        filter_data = static_cast<T*>(filter_dst_mem->get_data_handle());
+        if (provider_ != nullptr)
+          provider_->weights_mem_map[weightName] = filter_dst_mem;
```

Comment:
Is it thread safe?

Reply:
Good catch. Will add `lock_guard `to make this threads safe.